client/client.go

```diff
@@ -0,0 +1,604 @@
+			if err != nil {
+				return err
+			}
+			statusVM(ctx, &result)
+			return nil
+		})
+}
+
+// Run connects to the VM via SSH and runs the desired command
+func Run(ctx context.Context, instanceName, command string) error {
+	path, err := exec.LookPath("ssh")
+	if err != nil {
+		return fmt.Errorf("Unable to locate ssh binary")
+	}
+
+	var result types.InstanceDetails
+	err = issueCommand(ctx,
+		func(client *rpc.Client) (int, error) {
+			var id int
+			err := client.Call("ServerAPI.GetInstanceDetails", instanceName, &id)
+			return id, err
+		},
+		func(client *rpc.Client, id int) error {
+			err := client.Call("ServerAPI.GetInstanceDetailsResult", id, &result)
+			if err != nil {
+				return err
+			}
+			return nil
+		})
+
+	if err != nil {
+		return err
+	}
+
+	err = waitForSSH(ctx, &result, command != "")
+	if err != nil {
+		return err
+	}
+
+	args := []string{
+		path,
+		"-q", "-F", "/dev/null",
+		"-o", "UserKnownHostsFile=/dev/null",
+		"-o", "StrictHostKeyChecking=no",
+		"-o", "IdentitiesOnly=yes",
+		"-i", result.SSH.KeyPath,
+		result.VMSpec.HostIP.String(), "-p", strconv.Itoa(result.SSH.Port),
+	}
+
+	if command != "" {
+		args = append(args, command)
+	}
+
+	err = syscall.Exec(path, args, os.Environ())
+	return err
+}
+
+// Connect opens a shell to the VM via
+func Connect(ctx context.Context, instanceName string) error {
+	return Run(ctx, instanceName, "")
+}
+
+// Delete the VM
+func Delete(ctx context.Context, instanceName string) error {
+	return issueCommand(ctx,
+		func(client *rpc.Client) (int, error) {
+			var id int
+			err := client.Call("ServerAPI.Delete", instanceName, &id)
+			return id, err
+		},
+		func(client *rpc.Client, id int) error {
+			var result struct{}
+			return client.Call("ServerAPI.DeleteResult", id, &result)
+		})
+}
+
+// Instances provides information about all of the current instances
+func Instances(ctx context.Context) error {
+	var instances []string
+	err := issueCommand(ctx,
+		func(client *rpc.Client) (int, error) {
+			var id int
+			err := client.Call("ServerAPI.GetInstances", struct{}{}, &id)
+			return id, err
+		},
+		func(client *rpc.Client, id int) error {
+			return client.Call("ServerAPI.GetInstancesResult", id, &instances)
+		})
+
+	if err != nil {
+		return err
+	}
+
+	instanceDetails := make([]types.InstanceDetails, 0, len(instances))
+	for _, name := range instances {
+		_ = issueCommand(ctx,
+			func(client *rpc.Client) (int, error) {
+				var id int
+				err := client.Call("ServerAPI.GetInstanceDetails", name, &id)
+				return id, err
+			},
+			func(client *rpc.Client, id int) error {
+				var details types.InstanceDetails
+				err := client.Call("ServerAPI.GetInstanceDetailsResult", id, &details)
+				if err != nil {
+					return err
+				}
+				instanceDetails = append(instanceDetails, details)
+				return nil
+			})
+	}
+
+	if len(instanceDetails) == 0 {
+		return nil
+	}
+
+	w := new(tabwriter.Writer)
+	w.Init(os.Stdout, 0, 8, 1, '\t', 0)
```

Comment:
what's this? looks like magic numbers

Reply:
It's taken from the godoc example.  The numbers make sense if you read the godoc.